bufferserver/src/main/java/com/datatorrent/bufferserver/internal/DataList.java

```diff
@@ -405,7 +405,34 @@ public boolean resumeSuspendedClients(final int numberOfInMemBlockPermits)
 
   public boolean isMemoryBlockAvailable()
   {
-    return (storage == null) || (numberOfInMemBlockPermits.get() > 0);
+    return (numberOfInMemBlockPermits.get() > 0);
+  }
+
+  public boolean canAllocateMemoryBlock()
+  {
+    return ((storage == null) && !areSubscribersBehindByMax());
+  }
+
+  private boolean areSubscribersBehindByMax()
+  {
+    boolean behind = false;
+    if (backPressureEnabled) {
+      int count = 0;
+      synchronized (this) {
+        for (Block curr = last.prev; curr != null; curr = curr.prev, count++) {
+          // Since acquire happens before release, because of concurrency, in a corner case scenario we might still count a
+          // subscriber as being max behind when it is transitioning over to the next block but that is ok as it will only
+          // result in publisher blocking for some time and resuming
+          if (count >= (MAX_COUNT_OF_INMEM_BLOCKS - 2)) {
```

Comment:
It looks that there are two distinct stages of processing
- go back MAX_COUNT_OF_INMEM_BLOCKS - 1 blocks
- see if there is any block with a positive ref count
should it be expressed more explicitly? For example:
```
do {
  if (curr.refCount.get() != 0) {
    return true;
  }
  curr = curr.prev;
} while (curr != null);
return false;
```